### PR TITLE
Ledger state

### PR DESCRIPTION
**Pull request by jonjove**
These changes form the framework for the transition to the new data-layer. All of the objects that will make up the implementation of the data-layer are included. However, at this point the new data-layer is not yet used anywhere in stellar-core.

### CODE REVIEW

File: src/ledger/LedgerState.cpp [hunk 1]
```diff
@@ -0,0 +1,801 @@
...
+{
+    return (lhs.votes == rhs.votes) && (lhs.inflationDest == rhs.inflationDest);
+}
+
+LedgerStateRoot::LedgerStateRoot(Database& db) : mHasChild(false), mDb(db)
+{
+}
+
+void
+LedgerStateRoot::hasChild(bool child)
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: name of this method suggest that it returns bool, but it does not

File: src/ledger/LedgerState.cpp [hunk 2]
```diff
@@ -0,0 +1,801 @@
...
+Database&
+LedgerStateRoot::getDatabase()
+{
+    return mDb;
+}
+
+Database::EntryCache&
+LedgerStateRoot::getCache()
+{
+    // TODO(jonjove): Do we want to continue using the cache from Database?
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: +1 for moving cache here

File: src/ledger/LedgerState.cpp [hunk 3]
```diff
@@ -0,0 +1,801 @@
...
+    for (auto const& state : mParent->mState)
+    {
+        state.second->invalidate();
+    }
+    mParent->mLoadBestOfferContext.clear();
+}
+
+LedgerState::~LedgerState()
+{
+    if (mParent || mRoot)
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: is it possible to be true? I see only 2 constructors here and one always sets mParant and second one mRoot
maybe assert would be better?

File: src/ledger/LedgerState.cpp [hunk 4]
```diff
@@ -0,0 +1,801 @@
...
+LedgerEntryChanges
+LedgerState::getChanges()
+{
+    assert(mParent || mRoot);
+    assert(!mChild);
+
+    LedgerEntryChanges changes;
+    for (auto const& state : mState)
+    {
+        // TODO(jonjove): Should output meta if *entry == *previous?
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: probably: https://github.com/stellar/stellar-core/commit/fe955a910a43ae1a7c3ac744b579502b7565f01c

File: src/ledger/LedgerState.cpp [hunk 5]
```diff
@@ -0,0 +1,801 @@
...
+            }
+        }
+    }
+    return changes;
+}
+
+std::vector<LedgerEntry>
+LedgerState::getLiveEntries()
+{
+    assert(mRoot);
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: this looks like it is only allowed to be called by LedgerState just below root
what about moving that code to LedgerStateRoot?

File: src/ledger/LedgerState.cpp (hunk 3 above)
Comment (jonjove) [reply src/ledger/LedgerState.cpp]: This comes into play if `LedgerState::rollback` or `LedgerState::commit` has already been called. When either of those occur, we effectively detach the nested `LedgerState` from the rest of the chain. In this case, `mParent == mRoot == nullptr`. This prevents `LedgerState` from rolling back if it has already been handled. An analogous assert appears in `rollback` and `commit`.

File: src/ledger/LedgerState.cpp [hunk 6]
```diff
@@ -0,0 +1,801 @@
...
+
+void
+LedgerState::commit(std::function<void()> f)
+{
+    assert(mParent || mRoot);
+    assert(!mChild);
+
+    if (mParent)
+    {
+        assert(!f);
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: please use more descriptive name for that

File: src/ledger/LedgerState.cpp [hunk 7]
```diff
@@ -0,0 +1,801 @@
...
+        mergeStateIntoParent();
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: like those, those are great!

File: src/ledger/LedgerState.cpp [hunk 8]
```diff
@@ -0,0 +1,801 @@
...
+{
+    for (auto const& state : mState)
+    {
+        state.second->invalidate();
+        auto const& entry = state.second->ignoreInvalid().entry();
+
+        auto iter = mParent->mState.find(state.first);
+        if (iter == mParent->mState.end())
+        {
+            StateEntry ler(new LedgerEntryReference(entry, nullptr));
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: since C++ I don't like seeing naked new in code, what about make_shared?

File: src/ledger/LedgerState.cpp [hunk 9]
```diff
@@ -0,0 +1,801 @@
...
+
+    mHeader.reset();
+    mState.clear();
+    mLoadBestOfferContext.clear();
+}
+
+LedgerState::StateEntry
+LedgerState::create(LedgerEntry const& entry)
+{
+    assert(mParent || mRoot);
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: this set of assets appears everywhere, maybe promote it to a method?

File: src/ledger/LedgerState.cpp [hunk 10]
```diff
@@ -0,0 +1,801 @@
...
+LedgerState::StateHeader
+LedgerState::loadHeader()
+{
+    assert(mParent || mRoot);
+    assert(!mChild);
+
+    mHeader = loadHeaderHelper();
+    auto& header = mHeader->header();
+    ++header.ledgerSeq;
+    header.previousLedgerHash = sha256(xdr::xdr_to_opaque(header));
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: just wondering, will it use proper value of ledgerSeq as it was just incremented? seems strange to be that methods named "loadHeader" does modifications on header object

File: src/ledger/LedgerState.cpp (hunk 6 above)
Comment (jonjove) [reply src/ledger/LedgerState.cpp]: Definitely a good point. Renamed to `onCommitToDatabase`. This was added to allow some other SQL changes to be bundled into the SQL transaction that writes `LedgerState` to database. This will be used during ledger close as some things are written, for example, to `PersistentState` within the same SQL transaction.

File: src/ledger/LedgerState.cpp (hunk 8 above)
Comment (jonjove) [reply src/ledger/LedgerState.cpp]: `std::make_shared` can only be used when objects have public constructors (since otherwise `std::make_shared` does not have access to the constructor). But `LedgerEntryReference` and `LedgerHeaderReference` have private constructors as they can only be constructed by `LedgerState`. If you prefer it, we can promote this type of code into a method of its own. For example, something like `StateEntry makeSharedLedgerEntryReference(...)`. I think this is potentially a good idea, since it will mean that naked `new` will be in one place where we can have a clarifying comment explaining the necessity.

File: src/ledger/LedgerState.cpp (hunk 9 above)
Comment (jonjove) [reply src/ledger/LedgerState.cpp]: I considered this also (it was actually that way in the first draft of the code) but I wasn't sure it would actually be easier to follow in the future if that was a method. If the consensus is that it would be easier that way, I'm not opposed to changing it back.

File: src/ledger/LedgerState.cpp (hunk 10 above)
Comment (jonjove) [reply src/ledger/LedgerState.cpp]: Actually, this doesn't look quite right to me. I'm going to revisit this now. I hadn't written any tests for this function yet because I wasn't sure exactly what I wanted to test, but I'm going to write one now.

**Comment by vogel** [comment]
What do you think about extracting LedgerState classes to new directory? Either ledger/state or just ledgerstate?

File: src/ledger/LedgerState.cpp [hunk 11]
```diff
@@ -0,0 +1,801 @@
...
+            StateEntry ler(new LedgerEntryReference(entry, nullptr));
+            ler->invalidate();
+            mParent->mState[state.first] = ler;
+        }
+        else
+        {
+            auto const& previous =
+                iter->second->ignoreInvalid().previousEntry();
+            StateEntry ler(new LedgerEntryReference(entry, previous));
+            ler->invalidate();
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: This means that we don't want anyone to use that LedgerEntryReference anymore? Even if it was just created here and is 100% valid?

Do I understand this correctly? If so, please and comment and explain why is that needed. I think that this means we won't be able to use this entry in this LedgerState and one will have to create a new child LedgerState to use it?

File: src/ledger/LedgerState.cpp [hunk 12]
```diff
@@ -0,0 +1,801 @@
...
+}
+
+LedgerState::StateEntry
+LedgerState::createHelper(LedgerEntry const& entry, LedgerKey const& key)
+{
+    StateEntry ler;
+
+    auto iter = mState.find(key);
+    if (iter != mState.end())
+    {
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: when can this happen? it is unclear to me

File: src/ledger/LedgerState.cpp [hunk 13]
```diff
@@ -0,0 +1,801 @@
...
+}
+
+LedgerState::StateEntry
+LedgerState::loadHelper(LedgerKey const& key)
+{
+    StateEntry ler;
+
+    auto iter = mState.find(key);
+    if (iter != mState.end())
+    {
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: when can this happen? it is unclear to me

File: src/ledger/LedgerState.cpp [hunk 14]
```diff
@@ -0,0 +1,801 @@
...
+        auto const& header = mRoot->getPreviousHeader();
+        lhr = StateHeader(new LedgerHeaderReference(header, header));
+    }
+    return lhr;
+}
+
+LedgerState::StateEntry
+LedgerState::loadFromDatabase(LedgerKey const& key)
+{
+    assert(mRoot);
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: looks like this belong to LedgerStateRoot

File: src/ledger/LedgerState.cpp [hunk 15]
```diff
@@ -0,0 +1,801 @@
...
+    }
+    else
+    {
+        mRoot->getCache().put(cacheKey, nullptr);
+    }
+    return ler;
+}
+
+void
+LedgerState::storeInDatabase(StateEntry const& state)
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: looks like this belong to LedgerStateRoot

File: src/ledger/LedgerState.cpp [hunk 16]
```diff
@@ -0,0 +1,801 @@
...
+        mOffersLoadedFromDatabase += mFromDatabase.size();
+    }
+}
+
+LedgerState::StateEntry
+LedgerState::LoadBestOfferContext::loadBestOffer()
+{
+    if (mTop)
+    {
+        assert(!mTop->valid());
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: I find that counter intuitive that we are adding on purpose non-valid item to mInMemory here.
Maybe this is a problem with name of "valid" method?

File: src/ledger/LedgerState.cpp [hunk 17]
```diff
@@ -0,0 +1,801 @@
...
+        mFromDatabase = root->loadBestOffersFromDatabase(
+            5, mOffersLoadedFromDatabase, mSelling, mBuying);
+        mOffersLoadedFromDatabase += mFromDatabase.size();
+    }
+}
+
+LedgerState::StateEntry
+LedgerState::LoadBestOfferContext::loadBestOffer()
+{
+    if (mTop)
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: That inserts top value back into mInMemory. What is purpose of that? Won't it mean that each time the same value is returned?

File: src/ledger/LedgerState.cpp [hunk 18]
```diff
@@ -0,0 +1,801 @@
...
+            root = root->mParent;
+        }
+        mFromDatabase = root->loadBestOffersFromDatabase(
+            5, mOffersLoadedFromDatabase, mSelling, mBuying);
+        mOffersLoadedFromDatabase += mFromDatabase.size();
+    }
+}
+
+LedgerState::StateEntry
+LedgerState::LoadBestOfferContext::loadBestOffer()
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: I think maybe "popNextBestOffer" would be a better name, as it will be called multiple times and each time it should return different value. load preffix suggest that it should return the same value each time.

File: src/ledger/LedgerState.cpp [hunk 19]
```diff
@@ -0,0 +1,801 @@
...
+        return lhs.offerID < rhs.offerID;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+std::vector<InflationVotes>
+LedgerState::loadInflationWinners(size_t maxWinners, int64_t minBalance)
```
Comment (vogel) [inline src/ledger/LedgerState.cpp]: haha, this one is clever ;)

File: src/ledger/LedgerState.cpp (hunk 11 above)
Comment (jonjove) [reply src/ledger/LedgerState.cpp]: The `LedgerEntryReference` that is constructed here isn't exactly being created here. This is actually the process of copying the data from the child `LedgerState` to the parent `LedgerState`. So there was already a `LedgerEntryReference` that referenced this data, and it was already used as much as needed in the child `LedgerState` that is being committed.

I adopted the convention that any transaction-type operation (for example: commit, rollback, sub-transaction) would invalidate all existing `LedgerEntryReference`. Being invalid does not mean that you need to create a child `LedgerState` in order to use it, because it is valid to call `load` on the `LedgerKey` of an invalid `LedgerEntryReference` (see lines 372-397). Indeed, this is exactly what is done in the following example:
```
LedgerKey key = ...;
LedgerState lsParent(...);
LedgerState lsChild(lsParent);
someFunction(lsChild, key); // Some rarely accessed branches call lsChild.load(key)
lsChild.commit();
lsParent.load(key); 
```
Because the call to `lsChild.commit()` leaves every `LedgerEntryReference` in the invalid state, `lsParent.load(key)` will never fail with "Valid LedgerEntryReference already exists for this key" even if `someFunction` did not leave the `LedgerEntryReference` corresponding to `key` in the invalid state. 

It is worth noting that "re-loading" the invalidated `LedgerEntryReference` is a purely in-memory operation.

Let me know if this is still not clear, happy to explain more thoroughly or give more examples of different usage patterns (it might also be interesting to look at LedgerStateTests.cpp to see what kind of usage is possible).

File: src/ledger/LedgerState.cpp (hunk 12 above)
Comment (jonjove) [reply src/ledger/LedgerState.cpp]: Here is an example where this occurs:
```
LedgerKey key = ...;
LedgerState lsParent(...);
LedgerState lsChild(lsParent);
lsChild.load(key)->erase();
lsChild.commit();
lsParent.create(key);
```
Let me know if you need further explanation or examples on what is happening here, I'm happy to explain further.

File: src/ledger/LedgerState.cpp (hunk 13 above)
Comment (jonjove) [reply src/ledger/LedgerState.cpp]: The comment on line 226 gives an example of how this can happen.

File: src/ledger/LedgerState.cpp (hunk 18 above)
Comment (jonjove) [reply src/ledger/LedgerState.cpp]: The semantics here are actually more like "loadBestOffer" then "popNextBestOffer". `loadBestOffer` has analogous functionality to `std::priority_queue<...>::top()`. For example, the following test case should succeed:
```
Asset a1, a2 = ...;
LedgerState ls(...);
auto ler1 = ls.loadBestOffer(a1, a2);
auto le1 = *ler1->entry();
ler1->invalidate();
auto ler2 = ls.loadBestOffer(a1, a2);
auto le2 = *ler2->entry();
REQUIRE(le1 == le2);
```
If you want it to behave like "popNextBestOffer" than you would have to use `LedgerEntryReference::erase` like:
```
Asset a1, a2 = ...;
LedgerState ls(...);
auto ler1 = ls.loadBestOffer(a1, a2);
auto le1 = *ler1->entry();
ler1->erase(); // Now we have popped
ler1->invalidate();
auto ler2 = ls.loadBestOffer(a1, a2);
auto le2 = *ler2->entry();
REQUIRE(le1 != le2);
```
This functionality is natural to me, since you do not want to remove the best offer from the database every time you load it.

File: src/ledger/LedgerState.cpp (hunk 17 above)
Comment (jonjove) [reply src/ledger/LedgerState.cpp]: It does not mean that the same value is returned each time, although if the value is loaded and not modified or erased then it will be loaded again (as it should be, since it would still be the best offer at that point). The comment on line 686 gives an example where the offer loaded by `loadBestOffer` is erased, giving functionality similar to top followed by pop. Consider the case where the offer loaded by `loadBestOffer` is modified such that it is no longer the best offer. Then when it is pushed back into `mInMemory`, it will not necessarily be at the top of the heap.

File: src/ledger/LedgerState.cpp (hunk 16 above)
Comment (jonjove) [reply src/ledger/LedgerState.cpp]: I agree that this could indicate a problem with the name.

`LoadBestOfferContext` maintains the invariant that there is at most one valid `LedgerEntryReference` referring to an offer (with the correct assets), and if there is one then it is the one that was just returned by `loadBestOffer`. This means it is incorrect to call `loadBestOffer` if any `LedgerEntryReference` referring to an offer (with the correct assets) is valid.

Why is this required though? I think it would be possible to implement this without this constraint, but it would make it impossible to do the performance optimization which is done in `LoadBestOfferContext`. Specifically, if there could be multiple valid `LedgerEntryReferences` referring to offers (with the correct assets) then `LoadBestOfferContext::loadBestOffer` would have to re-process the entire `mInMemory` on each invocation. It is easy to see that this is the case. Any valid `LedgerEntryReference` can be modified at any time (and there is no callback to notify `LedgerState` that a modification has occurred). Suppose that I invoke `LoadBestOfferContext::loadBestOffer`, then modify many offers (with the correct assets), then call `LoadBestOfferContext::loadBestOffer` again. At this point `mInMemory` could be invalid (in the heap sense) at arbitrarily many locations, so it would need to be re-formed into a heap. But that would be no better than just iterating to find the best offer at every invocation. By making the restriction that if a `LedgerEntryReference` referring to an offer (with the correct assets) is only valid if it was just returned by `loadBestOffer` then we know that modifications can only occur at the top of the heap.

The consequence of this is that a sequence of calls to `loadBestOffer` can be efficient if you do not make any intervening calls to `LedgerState::load` or `LedgerState::create` that constructed a `LedgerEntryReference` referring to an offer (with the correct assets). If such an intervening call is made, then the `LoadBestOfferContext` is destroyed so that it has to be reconstructed at the next call to `loadBestOffer`, thereby maintaining the constraint.

File: src/ledger/LedgerState.cpp (hunk 10 above)
Comment (jonjove) [reply src/ledger/LedgerState.cpp]: Fixed.

**Comment by MonsieurNicolas** [comment]
This got super-seeded by #1600
